### PR TITLE
Subgraph

### PR DESCRIPTION
**Pull request by nevermore3**

### CODE REVIEW

File: src/validator/GetSubgraphValidator.cpp [hunk 1]
```diff
@@ -121,20 +121,40 @@ Status GetSubgraphValidator::validateBothInOutBound(BothInOutClause* out) {
...
+        }
+    }
+
+    // where *._dst udf_is_in listOfVids
+    ArgumentList* argList = new ArgumentList();
+    argList->addArgument(std::make_unique<EdgeDstIdExpression>(new std::string("*")));
+    for (const auto& i : listOfVids) {
+        argList->addArgument(std::make_unique<ConstantExpression>(std::move(i)));
+    }
+    auto* functionCall = new FunctionCallExpression(new std::string("udf_is_in"), argList);
```
Comment (CPWstatic) [inline src/validator/GetSubgraphValidator.cpp:146]: where *._dst IN collectVar{0}[0][0] OR *._dst IN collectVar{1}[0][0] OR ...

File: src/validator/GetSubgraphValidator.cpp [hunk 2]
```diff
@@ -144,74 +164,74 @@ Status GetSubgraphValidator::toPlan() {
...
+                           std::move(exprs),
+                           true /*subgraph not need duplicate*/);
+    gn1->setInputVar(projectVids->varName());
+
+    auto* filter = Filter::make(plan, gn1, buildFilterCondition());
+    filter->setInputVar(gn1->varName());
+    filter->setColNames({kVid});
+
+    // datacollect
     std::vector<std::string> collects = {gn1->varName()};
```
Comment (CPWstatic) [inline src/validator/GetSubgraphValidator.cpp:231]: {gn, gn1}

File: src/validator/GetSubgraphValidator.cpp [hunk 3]
```diff
@@ -144,74 +171,75 @@ Status GetSubgraphValidator::toPlan() {
...
+                           // TODO(jmq) add syntax like `BOTH *`, `OUT *` ...
+                           storage::cpp2::EdgeDirection::OUT_EDGE,   // FIXME: make direction right
+                           std::move(vertexProps),
+                           std::move(edgeProps),
+                           std::move(statProps),
+                           std::move(exprs),
+                           true /*subgraph not need duplicate*/);
+    gn1->setInputVar(projectVids->varName());
+
+    auto* filter = Filter::make(plan, gn1, buildFilterCondition(steps_));
```
Comment (CPWstatic) [inline src/validator/GetSubgraphValidator.cpp:234]: return qctx_->objPool()->add(buildFilterCondition(steps_));

Comment (nevermore3) [reply src/validator/GetSubgraphValidator.cpp:234]: done

File: tests/common/nebula_test_suite.py [hunk 4]
```diff
@@ -351,18 +352,97 @@ def to_value(self, col):
         elif isinstance(col, str):
             value.set_sVal(col.encode('utf-8'))
         elif isinstance(col, dict):
-            map_val = CommonTtypes.Map()
-            map_val.kvs = dict()
-            for key in col:
-                ok, temp = self.to_value(col[key])
+            if "type" not in col:
```
Comment (laura-ding) [inline tests/common/nebula_test_suite.py:355]: Maybe we need to modify the type to `__TYPE__`. When the result is map type and has the key is `type`.

File: tests/common/nebula_test_suite.py [hunk 5]
```diff
@@ -351,18 +352,97 @@ def to_value(self, col):
...
+        vertex = CommonTtypes.Vertex()
+        vertex.vid = bytes(v["vid"], encoding = 'utf-8')
+        tags = []
+        for tagVal in v["tags"]:
+            tag = CommonTtypes.Tag()
+            tag.name= bytes(tagVal["name"], encoding = 'utf-8')
+            props = dict()
+            for k, v in tagVal["props"].items():
+                value = CommonTtypes.Value()
+                if isinstance(v, int):
```
Comment (laura-ding) [inline tests/common/nebula_test_suite.py:384]: The type is not only int/float/string. you can use `self.to_value` directly.

File: tests/common/nebula_test_suite.py [hunk 6]
```diff
@@ -351,18 +352,97 @@ def to_value(self, col):
...
+    def convert_to_edge(self, e):
+        edge = CommonTtypes.Edge()
+        edge.src = bytes(e["src"], encoding = 'utf-8')
+        edge.dst = bytes(e["dst"], encoding = 'utf-8')
+        edge.type = e["type"]
+        edge.ranking = e["ranking"]
+        edge.name = bytes(e["name"], encoding = 'utf-8')
+        props = dict()
+        for k,v in e["props"].items():
+            value = CommonTtypes.Value()
```
Comment (laura-ding) [inline tests/common/nebula_test_suite.py:410]: ditto

File: tests/query/v1/test_get_subgraph.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+# attached with Common Clause Condition 1.0, found in the LICENSES directory.
+
+from tests.common.nebula_test_suite import NebulaTestSuite
+from tests.common.nebula_test_suite import T_EMPTY, T_NULL
+import pytest
+
+class TestGroupBy(NebulaTestSuite):
+    @classmethod
+    def prepare(self):
+        self.load_data()
```
Comment (laura-ding) [inline tests/query/v1/test_get_subgraph.py:15]: You can `self.use_nba()`. no need to insert the data again. tests have inserted the nba data at the beginning.

Comment (nevermore3) [reply tests/query/v1/test_get_subgraph.py]: done

File: tests/common/nebula_test_suite.py (hunk 6 above)
Comment (nevermore3) [reply tests/common/nebula_test_suite.py]: done

File: tests/common/nebula_test_suite.py (hunk 5 above)
Comment (nevermore3) [reply tests/common/nebula_test_suite.py]: done

File: tests/common/nebula_test_suite.py (hunk 4 above)
Comment (nevermore3) [reply tests/common/nebula_test_suite.py]: done

File: tests/common/nebula_test_suite.py [hunk 8]
```diff
@@ -425,6 +505,7 @@ def check_result(self, resp, expect, ignore_col: Set[int] = set(), is_regex = Fa
 
     @classmethod
     def check_out_of_order_result(self, resp, expect, ignore_col: Set[int] = set()):
+        print(resp)
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply tests/common/nebula_test_suite.py]: done

File: tests/common/nebula_test_suite.py [hunk 9]
```diff
@@ -377,7 +457,7 @@ def convert_expect(self, expect):
                 else:
                     ok, value = self.to_value(col)
                     if not ok:
-                        return ok, value
+                        return ok, value, 'jmq'
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply tests/common/nebula_test_suite.py]: done

File: src/validator/GetSubgraphValidator.cpp [hunk 10]
```diff
@@ -7,7 +7,7 @@
 #include "validator/GetSubgraphValidator.h"
 
 #include "common/expression/UnaryExpression.h"
-
+#include "common/expression/Expression.h"
```
[the remark this answers is not in the document]
Comment (nevermore3) [reply src/validator/GetSubgraphValidator.cpp:10]: done

File: tests/query/v1/test_get_subgraph.py [hunk 11]
```diff
@@ -0,0 +1,821 @@
...
+        expected_data = {
+            "column_names" : ['_vid', 'player.name', 'player.age', 'team.name', 'bachelor.name', 'bachelor.speciality'],
+            "rows" : [
+                ['Tim Duncan', 'Tim Duncan', 42, T_NULL, 'Tim Duncan', 'psychology'],
+                ['Tony Parker', 'Tony Parker', 36, T_NULL, T_NULL, T_NULL],
+                ['Spurs', T_NULL, T_NULL, 'Spurs', T_NULL, T_NULL]
+            ]
+        }
+        self.check_column_names(resp, expected_data["column_names"])
+        self.check_out_of_order_result(resp, expected_data["rows"])
```
Comment (CPWstatic) [inline tests/query/v1/test_get_subgraph.py:88]: We should return `vertex` in subgraph.

Comment (nevermore3) [reply tests/query/v1/test_get_subgraph.py:88]: will fix it in next PR

File: src/validator/TraversalValidator.cpp [hunk 12]
```diff
@@ -33,7 +33,11 @@ Status TraversalValidator::validateStarts(const VerticesClause* clause, Starts&
                    << "but was`" << type.value() << "'";
                 return Status::Error(ss.str());
             }
+
             starts.srcRef = src;
+            auto encode = src->encode();
+            auto decode = Expression::decode(encode);
+            startVidList_->add(decode.release());
```
Comment (CPWstatic) [inline src/validator/TraversalValidator.cpp:40]: `src` is a column, we should collect first.

**Review by yixinglu — APPROVED** [review]
LGTM

File: src/validator/TraversalValidator.cpp (hunk 12 above)
Comment (nevermore3) [reply src/validator/TraversalValidator.cpp:40]: done

**Review by CPWstatic — APPROVED** [review]

File: tests/query/v1/test_go.py [hunk 13]
```diff
@@ -1031,24 +1031,33 @@ def test_reversely_with_pipe(self):
         expected_data = {
             "column_names" : ["$^.team.name", "$$.player.name"],
             "rows" : [
-                ["Cavaliers", "Kyrie Irving"],
```
Comment (Shylock-Hg) [inline tests/query/v1/test_go.py:1034]: Why change this?

**Comment by CPWstatic** [comment]
Emm, we could add `collect_set` to improve the filter part.

File: tests/query/v2/test_get_subgraph.py [hunk 14]
```diff
@@ -0,0 +1,987 @@
...
+        expected_data = {
+            "column_names" : ["_vertices", "_edges"],
+            "rows" : [
+                [vertex1, edge1],
+                [vertex2, edge2]
+            ]
+        }
+
+        self.check_column_names(resp, expected_data["column_names"])
+        self.check_out_of_order_result(resp, expected_data["rows"])
```
Comment (laura-ding) [inline tests/query/v2/test_get_subgraph.py:987]: add line

File: tests/query/v2/test_get_subgraph.py [hunk 15]
```diff
@@ -0,0 +1,987 @@
...
+
+        stmt = '$a = GO FROM "Tim Duncan" OVER like YIELD $$.player.age AS ID; GET SUBGRAPH FROM $a.ID'
+        resp = self.execute_query(stmt)
+        self.check_resp_failed(resp)
+
+        stmt = '$a = GO FROM "Tim Duncan" OVER like YIELD like._src AS src; GET SUBGRAPH FROM $b.src'
+        resp = self.execute_query(stmt)
+        self.check_resp_failed(resp)
+
+        # stmt = 'GO FROM "Tim Duncan" OVER like YIELD $$._dst AS id, $^._src AS id | GET SUBGRAPH FROM $-.id'
```
Comment (laura-ding) [inline tests/query/v2/test_get_subgraph.py:42]: maybe want to test duplicate colname, modify to `YIELD like._dst AS id, like._src AS id`

File: tests/query/v2/test_get_subgraph.py [hunk 16]
```diff
@@ -0,0 +1,987 @@
...
+        # resp = self.execute_query(stmt)
+        # self.check_resp_failed(resp)
+
+        # stmt = '$a = GO FROM "Tim Duncan" OVER like YIELD $$._dst AS id, $^._src AS id; GET SUBGRAPH FROM $a.id'
```
Comment (laura-ding) [inline tests/query/v2/test_get_subgraph.py:46]: ditto

**Review by laura-ding — APPROVED** [review]

File: tests/query/v1/test_go.py (hunk 13 above)
Comment (nevermore3) [reply tests/query/v1/test_go.py:1034]: rank is missing from the  previous test case，after add rank field,  output of the statement before pipe are two 'Cavaliers'

**Pull request merged**